Add custom privacy request support for consent requests

galvana opened the pull request:
### Description Of Changes

Adding custom privacy request field support for consent requests. Introducing a new `disable_consent_identity_verification` that defaults to true or false based on the value of `subject_identity_verification_required` to keep the current behavior intact but allow admins to selectively disable identity verification for consent requests.

This PR only handles the collection and association of custom privacy request fields with a consent request. The usage of this new association is handled in fidesplus.

### Code Changes

* [ ] Updated `smoke_test.cy.ts` to assert the new fields in a consent request
* [ ] Updated the `ConsentConfig` and `ConsentRequestForm.tsx` to dynamically render and process `CustomPrivacyRequestFields`
* [ ] Updated the `/consent-request` endpoint to persist the custom privacy request fields in the database
* [ ] Database migration to add a reference between a `ConsentRequest` and `CustomPrivacyRequestField`s

There weren't any changes needed on the SaaS connector side of things, placeholder values (e.g. `<custom_privacy_request_fields.first_name>`) were already accounted for in the access/erasure phase of the custom privacy request work.

### Steps to Confirm

* [ ] E2E test must be completed in Fidesplus https://github.com/ethyca/fidesplus/pull/1285

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* [ ] Issue Requirements are Met
* [ ] Update `CHANGELOG.md`

galvana commented on src/fides/api/service/saas_request/override_implementations/appsflyer_request_overrides.py:
```diff
@@ -25,8 +25,8 @@ def appsflyer_user_read(
-    user_ids = input_data.get("user_id")
-    app_ids = input_data.get("app_id")
+    user_ids = input_data.get("user_id", [])
+    app_ids = input_data.get("app_id", [])
```
mypy fix for issue that made it to main

galvana commented on tests/ops/models/test_privacy_request.py:
```diff
@@ -977,9 +978,11 @@ class TestPrivacyRequestCustomFieldFunctions:
-    def test_cache_custom_privacy_request_fields(self):
-        CONFIG.execution.allow_custom_privacy_request_field_collection = True
-        CONFIG.execution.allow_custom_privacy_request_fields_in_request_execution = True
+    def test_cache_custom_privacy_request_fields(
```
Updating these old tests to use the new config fixtures

pattisdr reviewed (changes requested):
> This PR only handles the collection and association of custom privacy request fields with a consent request. The usage of this new association is handled in fidesplus.

Old Consent (config.json)  is entirely in Fides so the "usage" of the new association still needs to be handled here I believe!

pattisdr commented:
@galvana finished testing on the plus side, to echo here, Old Consent doesn't seem to work, New Consent seems to have the pieces in place.  

Are you sure there aren't any additional changes needed for the saas connectors - it all "just works" on the consent side?  I didn't verify this, just wanted to double check.

galvana commented on src/fides/api/alembic/migrations/versions/61a922702f4c_add_consent_request_relationship.py:
```diff
@@ -0,0 +1,34 @@
+        sa.Column("consent_request_id", sa.String(), nullable=True),
+    )
+    op.create_foreign_key(
+        None,
```
Renamed to `consent_request_id_fkey`

galvana commented on src/fides/api/api/v1/endpoints/consent_request_endpoints.py:
```diff
@@ -180,41 +181,53 @@ def create_consent_request(
     consent_request = ConsentRequest.create(db, data=consent_request_data)
 
-    if config_proxy.execution.subject_identity_verification_required:
+    consent_request.persist_custom_privacy_request_fields(
+        db=db, custom_privacy_request_fields=data.custom_privacy_request_fields
+    )
```
Thanks for outlining the flow! It looks like I was missing forwarding the consent request in `PATCH http://localhost:8080/api/v1/consent-request/<consent_request_id>/preferences`

galvana commented on src/fides/api/models/privacy_request.py:
```diff
@@ -1223,6 +1231,45 @@ def verify_identity(
+            for key, item in custom_privacy_request_fields.items():
+                if item.value:
+                    hashed_value = CustomPrivacyRequestField.hash_value(item.value)
+                    CustomPrivacyRequestField.create(
+                        db=db,
+                        data={
+                            "consent_request_id": self.id,
+                            "field_name": key,
+                            "field_label": item.label,
+                            "encrypted_value": {"value": item.value},
+                            "hashed_value": hashed_value,
+                        },
```
I added the custom fields to the `ConsentRequest` to minimize changes to the front-end. The front-end passes along the custom fields during the initial call to create a consent request but not for the subsequent privacy preference calls. Additionally, as you pointed out. We don't always create a `PrivacyRequest` for privacy preferences so we would need a separate location to store any custom fields.

I wasn't sure of the best way to handle this in the database, I don't like how there's two sets of custom fields. I was thinking of just moving the relationships between `ConsentRequest`, `PrivacyRequest`, and `CustomPrivacyRequestFields` to linking tables instead

`consent_request` <-> `consent_request_custom_fields` <-> `custom_privacy_request_fields`
`privacy_request` <-> `privacy_request_custom_fields` <-> `custom_privacy_request_fields`

But I wasn't sure if this was overkill. Open to hear what you think

pattisdr replied:
I still don't quite follow why we need custom privacy request field data if a privacy request isn't created because we're not using it for anything then? 

but assuming we do need it.  Can we link the same custom_privacy_request_field record to the consent_request_id first, and then if we come in and create a privacy request, edit those exisitng records with the privacy_request_id?

pattisdr replied:
@galvana can you look at this comment before merging? Let's make sure we're on the same page here

galvana replied:
In the original requirements for the custom privacy request fields work, we separated out the concepts of collecting custom fields, and using custom fields in a privacy request. In this case, if all we do is collect custom fields as part of a consent request, that is still a valid use case. With that in mind, I made the refactor you recommended above to only create one set of custom privacy request fields per consent/privacy request pair.

pattisdr replied:
OK got it thanks for clarifying that collecting custom fields is still a separate use case!

galvana commented on src/fides/config/config_proxy.py:
```diff
@@ -54,8 +54,27 @@ class ExecutionSettingsProxy(ConfigProxyBase):
+    disable_consent_identity_verification: bool
     require_manual_request_approval: bool
 
+    def __getattribute__(self, name: str) -> Any:
+        """
+        Overrides base __getattribute__ to provide a dynamic fallback for
+        'disable_consent_identity_verification'. The fallback is based on
+        'subject_identity_verification_required' only when no explicit value is provided,
+        preserving None for unset cases.
+        """
+        if name == "disable_consent_identity_verification":
+            value = super().__getattribute__("disable_consent_identity_verification")
+            if value is None:
+                subject_verification_required = super().__getattribute__(
+                    "subject_identity_verification_required"
+                )
+                return not subject_verification_required
+            return value
+
+        return super().__getattribute__(name)
+
```
That's correct! This was a tricky one that I got some help from Adam for. I originally had this fallback logic in the `ExecutionSettings` class but that didn't account for cases where `subject_identity_verification_required` was updated through the API. The behavior that I wanted was that as long as `disable_consent_identity_verification` isn't defined, it will continue to be derived from `subject_identity_verification_required`. Once a user sets a value (not `None`) then we will stop performing the fallback. The only way to get this to work is if we derive the fallback in the `ExecutionSettingsProxy` class.

galvana commented on src/fides/config/create.py:
```diff
@@ -192,6 +195,7 @@ def generate_config_docs(
     docs: str = build_config_header() + "\n".join(nested_settings_docs + object_docs)
 
     # Verify it is a valid Fides config file
+    print(docs)
```
Removed!

galvana commented:
@pattisdr, I added the missing pieces for the old consent workflow plus some other misc fixes. I was wondering why I was so off and I realized that my instructions for setting `FIDES_PRIVACY_CENTER__IS_OVERLAY_ENABLED` were wrong, those need to be set in `/clients` or `/clients/privacy-center`. I was setting it in the root fidesplus `.env` file thinking I was actually toggling between old and new consent 😵‍💫. I tested this locally and updated the tests for the old workflow.

pattisdr commented:
things are looking good in testing over here 👍 note that I just check as far as the privacy request gets created and is linked to this data.

pattisdr commented on src/fides/api/api/v1/endpoints/consent_request_endpoints.py:
```diff
@@ -407,6 +407,7 @@ def queue_privacy_request_to_propagate_consent_old_workflow(
                 identity=identity,
                 policy_key=policy,
                 consent_preferences=executable_consent_preferences,
+                consent_request_id=consent_request.id,
```
ah ok so probably a corresponding change needed plus side too?
